libyui/libyui-qt
# Prevent crash on [PrintScreen] of a main window (bsc#1157170)

## Trello

https://trello.com/c/enEoDUf9/1456-sles15-sp2-p1-1157170-pressing-print-screen-in-the-installer-makes-it-crash

## Bugzilla

https://bugzilla.suse.com/show_bug.cgi?id=1157170

## Problem

When using the Qt UI's built-in screenshot maker on the [PrintScreen] button on a main window (a window with `:Opt(:defaultsize)`), it caused a segfault. It did not segfault with a pop-up dialog.


## Cause

The changes for Qt 5.13 compatibility required different handling of this; it's now  no longer as easy as it used to be with the old (now deprecated) functions.

## Offending Commit

https://github.com/libyui/libyui-qt/commit/0cbda833d7056565a7ca6e111a8324851c0e2f34#diff-0ad7268eccfc51101fb060da6ee8c722

It turned out that the QWindow returned by `dialog->windowHandle()` could be 0 if it was a main window.

## Fix

Added an intermediate step that fetches the toplevel widget for that dialog. If it's a pop-up, it will just return itself. If it's a main window, it will return the widget that has an actual window.

## How to Test

Make sure that your desktop does not catch the [PrintScreen] key. In Xfce, check Settings -> Keyboard -> Application Shortcuts; you may have to reassign another key combination like Ctrl+PrintScreen to the existing action.

Start one of the UI examples that uses a main window: Events.rb, Wizard1.rb, MainWindow-popup.rb. Hit the [PrintScreen] button. You should see a file selection dialog that wants to save the screenshot to ~/.yast-screen-shots. Before the change, it crashed even before that; now it will just save the screenshot.

Also test with a pop-up dialog like HelloWorld.rb.
